include/rmm/device_uvector.hpp

```diff
@@ -119,6 +119,127 @@ class device_uvector {
+   * referenced by `v` after this function has returned.
+   *
+   * @note: This function incurs a host to device memcpy and should be used sparingly.
+   *
+   * Example:
+   * \code{cpp}
+   * rmm::device_uvector<int32_t> vec(100, stream);
+   *
+   * int v{42};
+   *
+   * // Copies 42 to element 0 on `stream` and synchronizes the stream
+   * vec.set_element(0, v, stream);
+   *
+   * // It is safe to destroy or modify `v`
+   * v = 13;
+   * \endcode
+   *
+   *
+   * @throws rmm::out_of_range exception if `element_index >= size()`
+   *
+   * @param element_index Index of the target element
+   * @param v The value to copy to the specified element
+   * @param s The stream on which to perform the copy
+   */
+  void set_element(std::size_t element_index, T const& v, cudaStream_t s)
+  {
+    RMM_EXPECTS(
+      element_index < size(), rmm::out_of_range, "Attempt to access out of bounds element.");
+    RMM_CUDA_TRY(cudaMemcpyAsync(element_ptr(element_index), &v, sizeof(v), cudaMemcpyDefault, s));
+    RMM_CUDA_TRY(cudaStreamSynchronize(s));
+  }
+
+  /**
+   * @brief Performs an *a*synchronous copy of `v` to the specified element in device memory.
+   *
+   * This function does _not_ synchronize stream `s` before returning. Therefore, the object
+   * referenced by `v` cannot be destroyed or modified until `stream` has been synchronized.
+   * Otherwise, behavior is undefined.
+   *
+   * @note: This function incurs a host to device memcpy and should be used sparingly.
+   *
+   * Example:
+   * \code{cpp}
+   * rmm::device_uvector<int32_t> vec(100, stream);
+   *
+   * int v{42};
+   *
+   * // Copies 42 to element 0 on `stream`. Does _not_ synchronize
+   * vec.set_element_async(0, v, stream);
+   * ...
+   * cudaStreamSynchronize(stream);
+   * // Synchronization is required before `v` can be modified
+   * v = 13;
+   * \endcode
+   *
+   * @throws rmm::out_of_range exception if `element_index >= size()`
+   *
+   * @param element_index Index of the target element
+   * @param v The value to copy to the specified element
+   * @param s The stream on which to perform the copy
+   */
+  void set_element_async(std::size_t element_index, value_type const& v, cudaStream_t s)
+  {
+    RMM_EXPECTS(
+      element_index < size(), rmm::out_of_range, "Attempt to access out of bounds element.");
+    RMM_CUDA_TRY(cudaMemcpyAsync(element_ptr(element_index), &v, sizeof(v), cudaMemcpyDefault, s));
+  }
+
+  /**
+   * @brief Returns the specified element from device memory
+   *
+   * @note: This function incurs a device to host memcpy and should be used sparingly.
+   *
+   * @throws rmm::out_of_range exception if `element_index >= size()`
+   *
+   * @param element_index Index of the desired element
+   * @param s The stream on which to perform the copy
+   * @return The value of the specified element
+   */
+  value_type get_element(std::size_t element_index, cudaStream_t s) const
```

Comment:
I think  maybe there should be a `get_element_async`. One may want to put other operations (independent kernels/copies) on the stream and only synchronize once `v` is really needed.

I'm thinking a common case would be to get a few disparate elements and only sync once.

```
int first{}, middle{}, last{};
v.get_element_async(&first, 0, s);
v.get_element_async(&middle, v.size()/2, s);
v.get_element_async(&last, v.size()-1, s);
cudaStreamSynchronize(s);
// access values
```

Reply:
That feels like an operation that would be better as a `thrust::copy`. If you're getting multiple elements, I think it's better to do it as a kernel than multiple memcopies.